modules/loggers/mod_log_config.c

```diff
@@ -1666,13 +1781,279 @@ static ap_log_writer *ap_log_set_writer(ap_log_writer *handle)
+            /* start sub object */
+            sub_log_object = apr_json_object_create(r->pool);
+
+            apr_json_object_set(log_object,
+                                attribute_name, APR_JSON_VALUE_STRING,
+                                sub_log_object,
+                                r->pool);
+
+            for (j = i; j < lfd->nelts; ++j) {
+                if(!items[j].tag) {
+                    continue;
+                }
+                if(strcmp(items[j].tag, items[i].tag) != 0) {
+                    i = j - 1;
+                    break;
+                }
+
+                attribute_name = items[j].arg;
+                if (!lfd->portions[j]) {
+                    apr_json_object_set(sub_log_object,
+                                        attribute_name, APR_JSON_VALUE_STRING,
+                                        json_null, r->pool);
+                }
+                else if (!*lfd->portions[j]) {
+                    apr_json_object_set(sub_log_object,
+                                        attribute_name, APR_JSON_VALUE_STRING,
+                                        json_empty_string,
+                                        r->pool);
+                }
+                else {
+                    attribute_value = lfd->portions[j];
+                    apr_json_object_set(sub_log_object,
+                                        attribute_name, APR_JSON_VALUE_STRING,
+                                        apr_json_string_create(r->pool, attribute_value,
+                                                               APR_JSON_VALUE_STRING),
+                                        r->pool);
+                }
+            }
+
+            continue;
+        }
+
+        if (!lfd->portions[i]) {
+            apr_json_object_set(log_object,
+                                attribute_name, APR_JSON_VALUE_STRING,
+                                json_null, r->pool);
+
+        }
+        else if (!*lfd->portions[i]) {
+            apr_json_object_set(log_object,
+                                attribute_name, APR_JSON_VALUE_STRING,
+                                json_empty_string,
+                                r->pool);
+        }
+        else {
+            attribute_value = lfd->portions[i];
+            apr_json_object_set(log_object,
+                                attribute_name, APR_JSON_VALUE_STRING,
+                                apr_json_string_create(r->pool, attribute_value,
+                                                       APR_JSON_VALUE_STRING),
+                                r->pool);
+
+        }
+    }
+
+    ba = apr_bucket_alloc_create(r->pool);
+    bb = apr_brigade_create(r->pool, ba);
+
+    /* encode into bucket brigade */
+    apr_json_encode(bb, NULL, NULL, log_object, APR_JSON_FLAGS_WHITESPACE, r->pool);
+    apr_brigade_puts(bb, NULL, NULL, APR_EOL_STR);
+
+    /* TODO: detect overflow nvec */
+    apr_brigade_to_iovec(bb, vec, &nvec);
```

Comment:
This dance makes me wonder whether we actually should use the apr_json API here (when it's available).
All we need is to build a json string (we won't work with or process the json values) so having to build an object value with key+value strings then serializing/encoding it in a bucket brigade then splitting the brigade in an iovec looks overkill[1] to me for the usage here.
If we need to keep ap_escape_logjson() anyway because apr-util < 1.7 or because it's usefull on its own (or elsewhere[2]) we possibly can keep the code simple(r) by just encoding the json string as we go so keep the `#else` ap_json_log_formatter() only.
If we keep this code we should avoid apr_brigade_to_iovec(), it seems indeed quite hard to use.. Using apr_brigade_pflatten() would create yet another buffer (possibly unbounded..) so maybe have our own helper like:
```
static apr_status_t brigade_to_iovec(apr_bucket_brigade *bb,
                                     struct iovec **pvec, int *nvec,
                                     apr_pool_t *p)
{
    apr_status_t rv = APR_SUCCESS;
    apr_array_header_t *a;
    const char *str;
    apr_size_t len;
    apr_bucket *b;
    int n = 0;

    for (b = APR_BRIGADE_FIRST(bb);
         b != APR_BRIGADE_SENTINEL(bb);
         b = APR_BUCKET_NEXT(b)) {
         n++;
    }
    a = apr_array_make(p, n, sizeof(struct iovec));
    for (b = APR_BRIGADE_FIRST(bb);
         b != APR_BRIGADE_SENTINEL(bb);
         b = APR_BUCKET_NEXT(b)) {
        rv = apr_bucket_read(b, &str, &len, APR_NONBLOCK_READ);
        if (rv != APR_SUCCESS) {
            break;
        }
        if (len) {
            struct iovec *iov = &APR_ARRAY_PUSH(a, struct iovec);
            iov->iov_base = (void *)str;
            iov->iov_len = len;
        }
    }
    *pvec = (void *)a->elts;
    *nvec = a->nelts;

    return rv;
}
```
?

[1] I'm not saying that the API is overkill since it's very well suited when one has to build or parse json and then work with the values (like mod_autht_jwk), but it's not what we do here (just produce a json string..). Btw I think that apr_json_encode() should take flags for what to do with non-utf8 strings (as-is/nocheck, replacement char, assume ISO-8859-1, error out, make tea..), likewise for apr_json_string_create[_ex]() because the closer to where the strings come from, the better (IMHO).

[2] mod_log_json is using libjansson (yet something else..) but with the `_nocheck` setters, ap_escape_logjson() could "enforce" valid json there too.